Security deployment guide (#2673)
zowe/docs-site: docs/user-guide/zowe-secure-setup.md

- Distributed Identity Federation
  - Okta
  - KeyCloak

## Transport Layer Security(TLS)

The TLS protocol should be used to ensure secure data-transport for all connections to API Mediation Layer services. 

### TLS requirements

* Java in version at least 8 sr6 fp25 is installed on the system.
* The following list shows the cipher suites that API ML services use.

```
TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA_POLY1305_SHA256
```

For more information, see the [TLS requirements in Zowe API ML requirements](../extend/extend-apiml/zowe-api-mediation-layer-security-overview#zowe-api-ml-tls-requirements).

### SAF keyring

You can choose to use a SAF keyring instead of keystore and truststore for storing certificates. It is more secure than PKCS12 files. And SAF keyring allows you to import existing or generate new certificates with Top Secret, ACF2, and RACF.

For details about SAF Keyring, see the documentation [here](../extend/extend-apiml/certificate-management-in-zowe-apiml.md).

## Authentication methods

The API Mediation Layer provides multiple methods which clients can use to authenticate.

### Authentication with JSON Web Tokens(JWT)

> When the client authenticates with the API ML, the client receives the JWT token in exchange. This token can be used for further authentication. The JWT secret that signs the JWT Token is an asymmetric private key that is generated during Zowe keystore configuration.

Reviewer: ```suggestion
When the user successfully authenticates with the API ML, the client receives a JWT token in exchange. This token can be used by the client to access REST services behind the API ML Gateway and also for subsequent user authentication. The access JWT Token is signed with the private key that is configured in the API ML Gateway certificate store, be it keystore or keyring.
```

> To utilize [Single-Sign-On (SSO)](../user-guide/systemrequirements-zos/#single-sign-on-sso), the Zowe API ML client needs to provide the access token to API services in the forma of the cookie `apimlAuthenticationToken`, or in the `Authorization: Bearer` HTTP header as described [here](https://github.com/zowe/sample-spring-boot-api-service/blob/master/zowe-rest-api-sample-spring/docs/api-client-authentication.md#authenticated-request).

Reviewer: ```suggestion
To utilize [Single-Sign-On (SSO)](../user-guide/systemrequirements-zos/#single-sign-on-sso), the Zowe API ML client needs to provide the access token to API services in the form of the cookie `apimlAuthenticationToken`, or in the `Authorization: Bearer` HTTP header as described [here](https://github.com/zowe/sample-spring-boot-api-service/blob/master/zowe-rest-api-sample-spring/docs/api-client-authentication.md#authenticated-request).
```

- Validation:
  - Use Gateway query endpoint to validate the token and retrieves the information associated with the token.
  - Use JWKS to validate.

- JWT token issuers
  - API ML - SAF, zOSMF LTPA (Lightweight Third-Party Authentication)
  - zOSMF - zOSMF JWT

Encoded JWT:
```
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c
```

> **Note** The endpoint is disabled by default. For more information, see [Enable JWT token endpoint](../user-guide/api-mediation/api-gateway-configuration#enable-jwt-token-refresh-endpoint).

Author: > Looks like this note is not in the correct place, as the document didn't mention the JWT refresh end-point till now. Or maybe the token refresh functionality needs to be described just above the note.
Will remove this note here in case of confusing users.

- The security administrator can configure an IDTDATA class profile to control how certain fields in an IDT are generated and how a specified IDT is validated in `RACROUTE REQUEST=VERIFY`. For details, see the ***Signed and Unsigned Identity Tokens*** and ***IDT Configuration*** subsections in [z/OS Security Server RACROUTE Macro Reference](https://www.ibm.com/docs/en/zos/2.4.0?topic=reference-activating-using-idta-parameter-in-racroute-requestverify).

## Multi-factor authentication (MFA)

Multi-factor authentication is provided by third-party products which Zowe is compatible with. CA AAM is recomended here. For details about multi-fator authentication, see [the MFA documentation here](../user-guide/mvd-configuration/#multi-factor-authentication-configuration).

### Certificate Authority Advanced Authentication Mainframe (CA AAM)

To add dynamic element to the authentication, you can configure the Certificate Authority Advanced Authentication Mainframe to enable multi-factor authentication. For details about CA AAM, see [the documentation](https://techdocs.broadcom.com/us/en/ca-mainframe-software/security/ca-advanced-authentication-mainframe/2-0.html).

> **Note:** MFA must work with [Single sign-on (SSO)](../user-guide/systemrequirements-zos/#single-sign-on-sso). Make sure that SSO is configured before you use MFA in Zowe.

Author: > What is meant by "MFA must work with SSO"?

I found it was mentioned in the [documentation](https://docs.zowe.org/stable/user-guide/systemrequirements-zos/#multi-factor-authentication-mfa) that SSO should be configured before using MFA in Zowe. If it is not accurate, what about removing "MFA must work with SSO" and only keeping "Make sure that SSO is configured before you use MFA in Zowe."?

## Authorization

The API ML can check for the authorization of the user on certain endpoints. Access to a SAF resource is checked with External Security Manager (ESM).

### SAF resource check

To verify the ownership of the SAF resource, you can use the following available providers:

- `native`(*default*) - on-platform
- `endpoint` - off-platform

**Example of the protected service**

`GET /gateway/{serviceId}/`

For detailed information, see the [SAF resource checking documentation](../user-guide/api-mediation/api-gateway-configuration/#saf-resource-checking).

## High Availability

To deploy Zowe high availability, you must set up the Parallel Sysplex® environment. A Parallel Sysplex is a collection of z/OS® systems that cooperatively use certain hardware and software components to achieve a high-availability workload processing environment.

### Sysplex architecture and configuration

Sysplex is required to make sure multiple Zowe instances can work together. Check [Configuring Sysplex for high availability](../user-guide/configure-sysplex) for more details.

To enable high availability when Zowe runs in Sysplex, you need to meet the following requirements:

> - Zowe instance with all services should be installed on every LPAR.

Reviewer: ```suggestion
- Zowe instance with should be installed on every LPAR.
- The API services must be registered to each Zowe instance.
```

- Shared File system should be created between LPARs in Sysplex. See [How to share file systems in a Sysplex](https://www.ibm.com/docs/en/zos/2.4.0?topic=planning-sharing-file-systems-in-sysplex). 
- z/OSMF High Availability mode should be configured. See [Configuring z/OSMF high availability in Sysplex](../user-guide/systemrequirements-zosmf-ha).

Instance on every LPAR is started.

**Configuration**

> The configuration for the specific instance is composed of the defaults in the main section and the overrides in the `haInstances` section.

Reviewer: ```suggestion
The configuration for the specific instance is composed of the defaults in the main section and the overrides in the `haInstances` section of the zowe.yaml configuration file.
```

In this section, <ha-instance> represents any Zowe high availability instance ID. Every instance has internal id and a section with overrides compared to the main configuration in the beginning of the `zowe.yaml` file. Check [Zowe YAML configuration reference](../appendix/zowe-yaml-configuration/#yaml-configurations---hainstances) for details.

### Caching service setup and configuration

Zowe uses the Caching Service to centralize the state data persistent in high availability (HA) mode. It can be used to share information between services.

If you are runnning the caching service on z/OS, there are three storage methods with their own characteristics:

- [VSAM](../user-guide/configure-caching-service-ha)

>   - Known by zOS engineers.

Reviewer: ```suggestion
  - Familiar to zOS engineers.
```

  - Slow.
- [Redis](../extend/extend-apiml/api-mediation-redis#redis-configuration)
  - Needs to run in Distributed world separately.
  - Good for Kubernetes deployment.
- [Infinispan (*recommended*)](../extend/extend-apiml/api-mediation-infinispan#infinispan-configuration)
  - Part of the Caching service.
  - Doesn’t need separate processes.
  - Highly performant.

## Observation

It is essential to have observation in the whole secure operation process. The service availability is visible to everyone. It’s easy to find performance issues with southbound services.

To do so, Zowe can easy integrate with the Alerting and Monitoring enterprise services such as ELK, Splunk, Grafana, etc and integrate the data as part of the Zero Trust Architecture approach with, for example SIEM (Security Incident Event Management). 

> ### Use Audit Log to observe

Reviewer 2: Remove the Audit Log please as it's vendor specific until we figure out whether and how we will open source the solution.

Author: Thanks for your comment Jakub! just want to double confirm if the whole `Observation` section should be removed as well.

Reviewer 2: No, the rest is still valid.

You can install the API Audit Log for Zowe API ML. The API `Audit Log` is an add-on to Zowe API Mediation Layer that provides unified auditing logs.

**Prerequisites**

- Brightside license is required.
- Elasticsearch, Logstash, and Kibana (ELK Stack) is configured. Using the API Audit Log also enables you to utilize ELK Stack for data visualizations.

For detailed information about the API Audit Log, see how to [Enable the API Audit Log](https://techdocs.broadcom.com/us/en/ca-mainframe-software/devops/ca-brightside/4-0/installing/install-brightside-extensions-and-z-os-components/enable-the-api-audit-log.html).

### Features

The API `Audit Log` can help you observe and analyze the data located on the mainframe, including suspicious activities and the state of the services.

> ## Distributed Identify Federation

Reviewer: This topic could be expanded to provide extended information.

Author: Hi Pavlin, thank you for your detailed comments and suggestions. Just wondering if you want this section to be another new page or just add more information here. Would you please leave some input or reference about the extended information you want here?

Author: > This topic could be expanded to provide extended information.

Hi Pavlin, thank you for your detailed comments and suggestions. Just wondering if you want this section to be another new page or just add more information here. Would you please leave some input or reference about the extended information you want here?

Reviewer: I think the topic deserves separate page (which don't have yet) and then could be introduced here with link to the detailed explanation.

Author: > I think the topic deserves separate page (which don't have yet) and then could be introduced here with link to the detailed explanation.

Will do. Before that, I'm wondering if the page will just cover the information like this draft, or you have further input to this page.
